pkg/preparer/setup.go

```diff
@@ -650,6 +653,83 @@ func getArtifactRegistry(preparerConfig *PreparerConfig) (artifact.Registry, err
 	return artifact.NewRegistry(url, fetcher, osversion.DefaultDetector), nil
 }
 
+func (p *Preparer) BuildRealityAtLaunch() error {
+	// check for pods on disk and not in intent
+	// insert kv pairs into reality if so
+	sub := p.Logger.SubLogger(nil)
+	// get set of pods in reality
+	realityResults, _, err := p.store.ListPods(consul.REALITY_TREE, p.node)
+	if err != nil {
+		sub.WithError(err).Errorln("Could not check reality: %s", err)
+		return err
+	}
+	realityMap := map[string]bool{}
+	for _, result := range realityResults {
+		realityMap[result.Manifest.ID().String()] = true
+	}
+	// get set of pods in intent
+	intentResults, _, err := p.store.ListPods(consul.INTENT_TREE, p.node)
+	if err != nil {
+		sub.WithError(err).Errorln("Could not check intent: %s", err)
+		return err
+	}
+	intentMap := map[string]bool{}
+	for _, result := range intentResults {
+		intentMap[result.Manifest.ID().String()] = true
+	}
+	// get set of pods installed on machine
+	manifestFilePaths, err := filepath.Glob(filepath.Join(p.podRoot, "*/current_manifest.yaml"))
+	if err != nil {
+		sub.WithError(err).Errorln("Could not check filepaths of pods installed on disk: %s", err)
+		return err
+	}
+	for _, fp := range manifestFilePaths {
+		home := strings.SplitN(fp, "/", 5)[3]
+		podUUID := types.HomeToPodUUID(home)
+		if podUUID != nil {
+			uniqueKey := types.PodUniqueKey(podUUID.String())
+			_, err := p.podStore.ReadPodFromIndex(podstore.PodIndex{PodKey: uniqueKey})
+			// no pod means the pod was unscheduled and also removed from the intent tree
+			if podstore.IsNoPod(err) {
+				ctx, cancelFunc := transaction.New(context.Background())
+				defer cancelFunc()
+				err := p.podStore.WriteRealityIndex(ctx, uniqueKey, p.node)
+				if err != nil {
+					sub.WithError(err).Errorln("Could not add 'write uuid index to reality store' to transaction: %s", err)
+					return err
+				}
+				ok, resp, err := transaction.Commit(ctx, p.client.KV())
+				if err != nil {
+					sub.WithError(err).Errorln("Could not write uuid index to reality store")
+					return err
+				}
+				if !ok {
+					err := util.Errorf("status record transaction rolled back: %s", transaction.TxnErrorsToString(resp.Errors))
+					sub.WithError(err).Errorln("Could not write uuid index to reality store")
+					return err
+				}
+			} else if err != nil {
+				sub.WithError(err).Errorln("Unexpected error reading pod: %s", err)
+				return err
+			}
+		} else if _, ok := intentMap[home]; !ok {
```

Comment:
Is this intent check even necessary? are you trying to prevent a case where the preparer was killed halfway through an install and therefore we shouldn't write reality yet?

Reply:
This check was here so that reality is only written to pods on disk not in intent. If it existed in intent then I thought it didn't need to be written to reality.

Reply:
yeah I think it's good to check intent, but i think it would be exceedingly rare for a pod to not be in reality but still be in intent, but probably good to be thorough